### ISSUE TITLE
Upgrade to use fflib causes differences

### ISSUE THREAD

**Issue by jondavis9898** [afawcett/declarative-lookup-rollup-summaries#209]
I think the move to use fflib was the right decision, moving away from classes that are essentially the same and taking advantage of ongoing enhancements available in the fflib project.

Unfortunately, this now presents an issue when both libraries (DLRS & FFLIB) are deployed to an org as files instead of packages.  For example, commit 3dd26dcc changed fflib_SObjectSelector.cls directly.  This could result in a constant chicken-egg problem between the two libraries.  In short, DLRS would have to be the priority project in an org and this could mean having to wait to have updates to fflib available until DLRS has caught up.

One thought for this is DLRS uses fflib as a package dependency instead of having the code directly in DLRS.  The shortcoming to this idea is that the change that was required in commit 3dd26dcc could not have been made to the fflib_SObjectSelector.cls because of the "privateness" of fflib (see https://github.com/financialforcedev/fflib-apex-common/issues/56).

Outside of always having everything installed as a package, thoughts on how this might be able to be solved or alternative solutions?

**Comment by afawcett** [comment]
In short the fact that unmanaged code all exists in the same namespace makes managing different versions (or customised versions) of shared libraries in an unmanaged scenario impossible. One day Salesforce may implement package scope as per Java. 

Making fflib packaged, requires the use of global access modifier, which then constrains the evolution of the library significantly. And in some way doesn't really support multiple versions either. 

The only real solution is to use managed packages for everything, as these have their own namespace and can contain their own copies of fflib. For me if i was developing in a production org customisation scenario, as in, in house development, i'd still be inclined to consider a managed package, where as of course most folks develop unmanaged in sandbox and promote unmanaged to production. Pros and cons i guess. 

Basically using managed packages for all code is the only way to isolate to this level, but has its overheads. DLRS is available as a managed package, so at least if you want your own unmanaged fflib version the two won't collide in this case.

**Comment by jondavis9898** [comment]
Thanks Andrew, appreciate the insight.  Hopefully Salesforce will enhance the platform in the future to make managing these scenario's a little easier (e.g. package scope, etc.).  For now, I tend to agree that managed packages even for in-house development is the "best" of the options.
